File: q0739/q0739.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];

        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = temperatures.length - 1; i >= 0; i--) {
            int currTemp = temperatures[i];
            // maintain the stack to be monotonic decreasing
            while (!stack.isEmpty() && currTemp >= temperatures[stack.peek()])
                stack.pop();

            // update the res if there is anything left on the stack
            if (!stack.isEmpty())
                res[i] = stack.peek() - i;

            // push the currTemp index onto the stack for future tracking
            stack.push(i);
        }

        return res;
    }
}

File: q0739/q0739.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        res = [0] * len(temperatures)
        
        # Use monotonic decreasing stack to keep track of the temperatures
        monoStack = []

        for i in range(len(temperatures)-1, -1, -1):
            currTemperature = temperatures[i]
            # maintain the stack to be monotonic decreasing
            while len(monoStack) > 0 and currTemperature >= temperatures[monoStack[-1]]: 
                monoStack.pop()
            # update the res if there is anything left on the stack
            if len(monoStack) != 0: 
                res[i] = monoStack[-1] - i
            # push the currTemp index onto the stack for future tracking
            monoStack.append(i)
        
        return res